crates/virtio-queue/src/descriptor_utils.rs

```diff
@@ -0,0 +1,939 @@
+        let total_bytes_consumed =
+            self.bytes_consumed
+                .checked_add(bytes_consumed)
+                .ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::InvalidData, Error::DescriptorChainOverflow)
+                })?;
+
+        let mut rem = bytes_consumed;
+        while let Some(vs) = self.buffers.pop_front() {
+            if rem < vs.len() {
+                // Split the slice and push the remainder back into the buffer list. Safe because we
+                // know that `rem` is not out of bounds due to the check and we checked the bounds
+                // on `vs` when we added it to the buffer list.
+                self.buffers.push_front(vs.offset(rem).unwrap());
+                break;
+            }
+
+            // No need for checked math because we know that `vs.size() <= rem`.
+            rem -= vs.len();
+        }
+
+        self.bytes_consumed = total_bytes_consumed;
+
+        Ok(bytes_consumed)
+    }
+
+    fn split_at(&mut self, offset: usize) -> Result<DescriptorChainConsumer<'a, B>> {
+        let mut rem = offset;
+        let pos = self.buffers.iter().position(|vs| {
+            if rem < vs.len() {
+                true
+            } else {
+                rem -= vs.len();
+                false
+            }
+        });
+
+        if let Some(at) = pos {
+            let mut other = self.buffers.split_off(at);
+
+            if rem > 0 {
+                // There must be at least one element in `other` because we checked
+                // its `size` value in the call to `position` above.
+                let front = other.pop_front().expect("empty VecDeque after split");
+                self.buffers
+                    .push_back(front.subslice(0, rem).map_err(Error::VolatileMemoryError)?);
+                other.push_front(front.offset(rem).map_err(Error::VolatileMemoryError)?);
+            }
+
+            Ok(DescriptorChainConsumer {
+                buffers: other,
+                bytes_consumed: 0,
+            })
+        } else if rem == 0 {
+            Ok(DescriptorChainConsumer {
+                buffers: VecDeque::new(),
+                bytes_consumed: 0,
+            })
+        } else {
+            Err(Error::SplitOutOfBounds(offset))
+        }
+    }
+}
+
+/// Provides high-level interface over the sequence of memory regions
+/// defined by readable descriptors in the descriptor chain.
+///
+/// Note that virtio spec requires driver to place any device-writable
+/// descriptors after any device-readable descriptors (2.6.4.2 in Virtio Spec v1.1).
+/// Reader will skip iterating over descriptor chain when first writable
+/// descriptor is encountered.
+#[derive(Clone)]
+pub struct Reader<'a, B = ()> {
+    buffer: DescriptorChainConsumer<'a, B>,
+}
+
+impl<'a, B: Bitmap + BitmapSlice + 'static> Reader<'a, B> {
+    /// Construct a new Reader wrapper over `desc_chain`.
+    pub fn new<M>(
+        mem: &'a GuestMemoryMmap<B>,
```

Comment:
This should ideally just depend on the GuestMemory trait. Unfortunately, that becomes a bit messy with querying the regions later... I got some ideas to clean that up, but we probably need to do it step by step...

Reply:
Yep, but ideally, we could stay generic over the concrete GuestMemory implementation. So it should probably be a `GuestMemory` here (`GuestMemoryMmap` is one implementation of `GuestMemory`). However, it looks like the generics in vm-memory are a bit weird. I _think_ I found a way to clean that up. It mostly seems to be related to some weird higher-rank lifetime generic constraints. But I need a bit more time to figure out the best solution.

Reply:
@Ablu, after discussing with @stefano-garzarella, if using GuestMemory is not possible, do you think it is correct to define these helpers in DescriptorChain?

Reply:
@MatiasVara do you mean the helpers to create the reader or the whole `descriptor_utils.rs` file?

Reply:
yes, i.e., the writer and reader which are defined in the context of DescriptorChain. Shall we remove them from there?

Reply:
> yes, i.e., the writer and reader which are defined in the context of DescriptorChain. Shall we remove them from there?

I got no opinion on that. I am fine with either. Will leave it to CODEOWNERs to comment.

Reply:
After a second look, I think it's fine for now (if the maintainers agree) to leave them in DescriptorChain.